plugin/settings/settings_storage.py

```diff
@@ -0,0 +1,209 @@
+""" Holds a class that encapsulates plugin settings
+
+Attributes:
+    log (logging.Logger): logger for this module
+"""
+import sublime
+import logging
+import re
+
+from os import path
+
+from ..tools import Tools
+
+log = logging.getLogger(__name__)
+log.debug(" reloading module %s", __name__)
+
+
+class Wildcards:
+    """ Enum class of supported wildcards
+
+    Attributes:
+        CLANG_VERSION (str): Description
+        PROJECT_NAME (str): Description
+        PROJECT_PATH (str): Description
+    """
+    PROJECT_PATH = "$project_base_path"
+    PROJECT_NAME = "$project_name"
+    CLANG_VERSION = "$clang_version"
+
+
+class SettingsStorage:
+    """ A class that stores all loaded settings
+
+    Attributes:
+        CMAKE_PRIORITIES (str[]): possible cmake properties
+        max_tu_age (int): maximum TU age in seconds
+        NAMES_ENUM (str[]): all supported settings names
+        PREFIXES (str[]): setting prefixes supported by this plugin
+    """
+    CMAKE_PRIORITIES = ["ask", "merge", "overwrite", "keep_old"]
+    PREFIXES = ["ecc_", "easy_clang_complete_"]
+
+    _wildcard_values = {
+        Wildcards.PROJECT_PATH: "",
+        Wildcards.PROJECT_NAME: "",
+        Wildcards.CLANG_VERSION: ""
+    }
+
+    # refer to Preferences.sublime-settings for usage explanation
+    NAMES_ENUM = [
+        "autocomplete_all",
+        "c_flags",
+        "clang_binary",
+        "cmake_flags_priority",
+        "cmake_prefix_paths",
+        "common_flags",
+        "cpp_flags",
+        "errors_on_save",
+        "generate_flags_with_cmake",
+        "hide_default_completions",
+        "include_file_folder",
+        "include_file_parent_folder",
+        "max_tu_age",
+        "search_clang_complete_file",
+        "triggers",
+        "use_libclang",
+        "verbose",
+    ]
+
+    def __init__(self, settings_handle):
+        """ Initialize settings storage with default settings handle
+
+        Args:
+            settings_handle (sublime.Settings): handle to sublime settings
+        """
+        self.__load_vars_from_settings(settings_handle, project_specific=False)
+
+    def update_from_view(self, view):
+        """ Update from view using view-specific settings
+
+        Args:
+            view (sublime.View): current view
+        """
+        self.__populate_common_flags(view)
+        self.__load_vars_from_settings(view.settings(), project_specific=True)
+
+    def is_valid(self):
+        """ Check settings validity. If any of the settings is None the settings
+        are not valid.
+
+        Returns:
+            bool: validity of settings
+        """
+        for key, value in self.__dict__.items():
```

Comment:
I would prefer to leave it this way as there can be some variables like `project_path` that are created but are not part of the enum. Now they cannot be `None`, but if new ones get introduced at some point we will automatically catch if they are invalid.